### ISSUE TITLE
LR keeps growing instead of shrinking

### ISSUE THREAD

**Issue by achaiah** [JianGoForIt/YellowFin_Pytorch#19]
Hi,

I'm running into a situation where YellowFin keeps adjusting the LR upward instead of decaying it downward - which, of course, prevents the network from converging.  Any idea why it would be happening?  Thanks!

I initialize YF like so: `optimizer = YFOptimizer(net.parameters(), lr=train_args['lr'])`

And the output is...

```
[epoch 4], [iter 20 / 123], [train main loss 0.15482], [lr 0.057262]
[epoch 4], [iter 40 / 123], [train main loss 0.14807], [lr 0.058468]
[epoch 4], [iter 60 / 123], [train main loss 0.14976], [lr 0.059635]
[epoch 4], [iter 80 / 123], [train main loss 0.14867], [lr 0.060768]
[epoch 4], [iter 100 / 123], [train main loss 0.14935], [lr 0.061881]
[epoch 4], [iter 120 / 123], [train main loss 0.14653], [lr 0.062980]

----------------------------------------------------------------------------------------------
[epoch 5], [iter 20 / 123], [train main loss 0.14709], [lr 0.064228]
[epoch 5], [iter 40 / 123], [train main loss 0.14188], [lr 0.065275]
[epoch 5], [iter 60 / 123], [train main loss 0.16187], [lr 0.066297]
[epoch 5], [iter 80 / 123], [train main loss 0.15231], [lr 0.067289]
[epoch 5], [iter 100 / 123], [train main loss 0.15639], [lr 0.068227]
[epoch 5], [iter 120 / 123], [train main loss 0.15515], [lr 0.069117]

----------------------------------------------------------------------------------------------
[epoch 6], [iter 20 / 123], [train main loss 0.13752], [lr 0.070135]
[epoch 6], [iter 40 / 123], [train main loss 0.13210], [lr 0.071002]
[epoch 6], [iter 60 / 123], [train main loss 0.13821], [lr 0.071850]
[epoch 6], [iter 80 / 123], [train main loss 0.13456], [lr 0.072690]
[epoch 6], [iter 100 / 123], [train main loss 0.13225], [lr 0.073533]
[epoch 6], [iter 120 / 123], [train main loss 0.13367], [lr 0.074379]
```

**Comment by JianGoForIt** [comment]
Thank you for the report! This behavior could be desirable in some cases, but it also could signify some issue. We are very interested in understanding it. Could you share some more information about your model?

**Comment by achaiah** [comment]
In my case the behavior is definitely not right as the LR never shrinks and only grows with time to astronomical numbers. I was looking to use YF on a segmentation task with UNet but ran into this problem.

**Comment by achaiah** [comment]
Just wanted to comment here again that after pulling down the newest Yellowfin code, all of my projects (that used to work well) have increasing learning rates.  Are you sure there isn't a sign somewhere that got flipped?  Seems very suspicious.....

**Comment by JianGoForIt** [comment]
Hi @achaiah,

Thanks for the feedback.

Maybe just a couple of comments which may help in your situation:

1. Could you make sure that the loss function are normalized as discussed on the repo Readme?

2. You may consider using YFOptimizer.set_lr_factor() to finer control the learning rate if that is necessary.

3. Now the the lr argument in YFOptimizer initialization is just placeholder for backward compatibility. It is not used as currently the optimizer does not have any initial lr and mu. set_lr_factor() is the tool you might want consider if you want to make the learning rate smaller. :)

Best

**Comment by JianGoForIt** [comment]
Hi @achaiah 

btw, in our new version, we have a slow start phase where we increase lr gradually at the begining to prevent some unstable behaviors we have before at the beginning of training. Not sure if your observation is related to this fact.

**Comment by mitliagkas** [comment]
Dear @achaiah, thank you for trying the latest version! Would you be able to share your model with us? Please feel free to reach out to us over email. Thanks!

**Comment by mitliagkas** [comment]
Dear @achaiah, thank you for trying the latest version! Would you be able to share your model with us? Please feel free to reach out to us over email. Thanks!

**Comment by achaiah** [comment]
Hi, for these tests I'm using the standard resnet18 model from pytorch model zoo. My dataset is the 17flowers dataset from [here](http://www.robots.ox.ac.uk/~vgg/data/flowers/17/).

Here is a log of training for 15 epochs on 3 GPUs.  You can see that at no point is the LR going down :(

```
Epoch 1/15: 19 batches [00:06,  1.15 batches/s, loss=2.85, val_loss=2.47, learn_rates=[0.0001]]
Epoch 2/15: 19 batches [00:05,  1.46 batches/s, loss=2.77, val_loss=2.26, learn_rates=[0.00089726455786560229]]
Epoch 3/15: 19 batches [00:05,  1.46 batches/s, loss=2.45, val_loss=2, learn_rates=[0.0018215616808968829]]
Epoch 4/15: 19 batches [00:05,  1.44 batches/s, loss=2.21, val_loss=1.68, learn_rates=[0.002770575491151625]]
Epoch 5/15: 19 batches [00:05,  1.36 batches/s, loss=1.74, val_loss=1.34, learn_rates=[0.0037469889384629694]]
Epoch 6/15: 19 batches [00:05,  1.44 batches/s, loss=1.37, val_loss=0.993, learn_rates=[0.0047843377388533601]]
Epoch 7/15: 19 batches [00:05,  1.43 batches/s, loss=0.858, val_loss=0.711, learn_rates=[0.0059166850489839502]]
Epoch 8/15: 19 batches [00:05,  1.44 batches/s, loss=0.909, val_loss=0.516, learn_rates=[0.0071631171758583731]]
Epoch 9/15: 19 batches [00:05,  1.42 batches/s, loss=0.578, val_loss=0.411, learn_rates=[0.0084926871001262134]]
Epoch 10/15: 19 batches [00:05,  1.35 batches/s, loss=0.461, val_loss=0.328, learn_rates=[0.0099328385449176972]]
Epoch 11/15: 19 batches [00:05,  1.38 batches/s, loss=0.362, val_loss=0.276, learn_rates=[0.011464789605204357]]
Epoch 12/15: 19 batches [00:05,  1.45 batches/s, loss=0.508, val_loss=0.239, learn_rates=[0.01305976302907128]]
Epoch 13/15: 19 batches [00:05,  1.42 batches/s, loss=0.277, val_loss=0.231, learn_rates=[0.013643691954522828]]
Epoch 14/15: 19 batches [00:05,  1.45 batches/s, loss=0.438, val_loss=0.188, learn_rates=[0.014101111463146802]]
Epoch 15/15: 19 batches [00:05,  1.45 batches/s, loss=0.223, val_loss=0.176, learn_rates=[0.014548990741424579]]
```

**Comment by mitliagkas** [comment]
Thank you for the log!

What I see here is that the validation loss is consistently decreasing, which is desirable behavior! What happens after 15 epochs? Do you see any instabilities? Does performance, as measured by the validation loss, deteriorate after some point?

Having an increasing LR is not necessarily a problem (I can elaborate) as long as it doesn't explode causing instabilities. Finally, using the validation loss to do early stopping is common practice and a good idea.

Looking forward to your answers and comments!

**Comment by JianGoForIt** [comment]
Please let me know if the discussion is active again.

**Comment by mitliagkas** [comment]
Thanks for the update, @achaiah; this wasn't mentioned the last time we discussed this (val loss was behaving nicely in your last listing). 

Any chance you can share your model? We'd love to study it!!
